sklearn/utils/_weight_vector.pyx.tp

```diff
@@ -145,27 +174,30 @@ cdef class WeightVector(object):
 
         Returns
         -------
-        innerprod : double
+        innerprod : {{c_type}}
             The inner product of ``x`` and ``w``.
         """
         cdef int j
         cdef int idx
-        cdef double innerprod = 0.0
-        cdef double* w_data_ptr = self.w_data_ptr
+        cdef {{c_type}} innerprod = 0.0
+        cdef {{c_type}}* w_data_ptr = self.w_data_ptr
         for j in range(xnnz):
             idx = x_ind_ptr[j]
             innerprod += w_data_ptr[idx] * x_data_ptr[j]
         innerprod *= self.wscale
         return innerprod
 
-    cdef void scale(self, double c) nogil:
+    cdef void scale(self, {{c_type}} c) nogil:
         """Scales the weight vector by a constant ``c``.
 
         It updates ``wscale`` and ``sq_norm``. If ``wscale`` gets too
         small we call ``reset_swcale``."""
         self.wscale *= c
         self.sq_norm *= (c * c)
-        if self.wscale < 1e-9:
+
+        cdef {{c_type}} threshold = {{reset_wscale_threshold}}
+
+        if self.wscale < threshold:
```

Comment:
Wouldn't Cython and C compilers be able to do the right thing with the simpler:

```suggestion
        if self.wscale < {{reset_wscale_threshold}}:
```

Reply:
In both cases, this seems to map to plain C:
```c
  /* "wv.pyx":195
 * 
 * 
 *         if self.wscale < 1e-9:             # <<<<<<<<<<<<<<
 *             self.reset_wscale()
 * 
 */
  __pyx_t_1 = ((__pyx_v_self->wscale < 1e-9) != 0);
  if (__pyx_t_1) {
```

<details>
<summary>File used to test</summary>

```bash
cythonize -3 wv.pyx
```

```python
# wv.pyx
# Tested compiling for both `float` and `double`
 
# cython: cdivision=True
# cython: boundscheck=False
# cython: wraparound=False

cimport cython
from libc.limits cimport INT_MAX
from libc.math cimport sqrt
import numpy as np
cimport numpy as np

np.import_array()


cdef class WeightVector64(object):
    """Dense vector represented by a scalar and a numpy array.

    The class provides methods to ``add`` a sparse vector
    and scale the vector.
    Representing a vector explicitly as a scalar times a
    vector allows for efficient scaling operations.

    Attributes
    ----------
    w : ndarray, dtype=float, order='C'
        The numpy array which backs the weight vector.
    aw : ndarray, dtype=float, order='C'
        The numpy array which backs the average_weight vector.
    w_data_ptr : float*
        A pointer to the data of the numpy array.
    wscale : float
        The scale of the vector.
    n_features : int
        The number of features (= dimensionality of ``w``).
    sq_norm : float
        The squared norm of ``w``.
    """
    cdef readonly float[::1] w
    cdef readonly float[::1] aw
    cdef float *w_data_ptr
    cdef float *aw_data_ptr
    cdef float wscale
    cdef float average_a
    cdef float average_b
    cdef int n_features
    cdef float sq_norm

    def __cinit__(self,
                  float[::1] w,
                  float[::1] aw):

        if w.shape[0] > INT_MAX:
            raise ValueError("More than %d features not supported; got %d."
                             % (INT_MAX, w.shape[0]))
        self.w = w
        self.w_data_ptr = &w[0]
        self.wscale = 1.0
        self.n_features = w.shape[0]

        self.aw = aw
        if self.aw is not None:
            self.aw_data_ptr = &aw[0]
            self.average_a = 0.0
            self.average_b = 1.0

    cdef void add(self, float *x_data_ptr, int *x_ind_ptr, int xnnz,
                  float c) nogil:
        """Scales sample x by constant c and adds it to the weight vector.

        This operation updates ``sq_norm``.

        Parameters
        ----------
        x_data_ptr : float*
            The array which holds the feature values of ``x``.
        x_ind_ptr : np.intc*
            The array which holds the feature indices of ``x``.
        xnnz : int
            The number of non-zero features of ``x``.
        c : float
            The scaling constant for the example.
        """
        cdef int j
        cdef int idx
        cdef float val
        cdef float innerprod = 0.0
        cdef float xsqnorm = 0.0

        # the next two lines save a factor of 2!
        cdef float wscale = self.wscale
        cdef float* w_data_ptr = self.w_data_ptr

        for j in range(xnnz):
            idx = x_ind_ptr[j]
            val = x_data_ptr[j]
            innerprod += (w_data_ptr[idx] * val)
            xsqnorm += (val * val)
            w_data_ptr[idx] += val * (c / wscale)

        self.sq_norm += (xsqnorm * c * c) + (2.0 * innerprod * wscale * c)

    # Update the average weights according to the sparse trick defined
    # here: https://research.microsoft.com/pubs/192769/tricks-2012.pdf
    # by Leon Bottou
    cdef void add_average(self, float *x_data_ptr, int *x_ind_ptr, int xnnz,
                          float c, float num_iter) nogil:
        """Updates the average weight vector.

        Parameters
        ----------
        x_data_ptr : float*
            The array which holds the feature values of ``x``.
        x_ind_ptr : np.intc*
            The array which holds the feature indices of ``x``.
        xnnz : int
            The number of non-zero features of ``x``.
        c : float
            The scaling constant for the example.
        num_iter : float
            The total number of iterations.
        """
        cdef int j
        cdef int idx
        cdef float val
        cdef float mu = 1.0 / num_iter
        cdef float average_a = self.average_a
        cdef float wscale = self.wscale
        cdef float* aw_data_ptr = self.aw_data_ptr

        for j in range(xnnz):
            idx = x_ind_ptr[j]
            val = x_data_ptr[j]
            aw_data_ptr[idx] += (self.average_a * val * (-c / wscale))

        # Once the sample has been processed
        # update the average_a and average_b
        if num_iter > 1:
            self.average_b /= (1.0 - mu)
        self.average_a += mu * self.average_b * wscale

    cdef float dot(self, float *x_data_ptr, int *x_ind_ptr,
                    int xnnz) nogil:
        """Computes the dot product of a sample x and the weight vector.

        Parameters
        ----------
        x_data_ptr : float*
            The array which holds the feature values of ``x``.
        x_ind_ptr : np.intc*
            The array which holds the feature indices of ``x``.
        xnnz : int
            The number of non-zero features of ``x`` (length of x_ind_ptr).

        Returns
        -------
        innerprod : float
            The inner product of ``x`` and ``w``.
        """
        cdef int j
        cdef int idx
        cdef float innerprod = 0.0
        cdef float* w_data_ptr = self.w_data_ptr
        for j in range(xnnz):
            idx = x_ind_ptr[j]
            innerprod += w_data_ptr[idx] * x_data_ptr[j]
        innerprod *= self.wscale
        return innerprod

    cdef void scale(self, float c) nogil:
        """Scales the weight vector by a constant ``c``.

        It updates ``wscale`` and ``sq_norm``. If ``wscale`` gets too
        small we call ``reset_swcale``."""
        self.wscale *= c
        self.sq_norm *= (c * c)


        if self.wscale < 1e-9:
            self.reset_wscale()

    cdef void reset_wscale(self) nogil:
        """Scales each coef of ``w`` by ``wscale`` and resets it to 1. """
        if self.aw_data_ptr != NULL:
            self.average_a = 0.0
            self.average_b = 1.0

        self.wscale = 1.0

    cdef float norm(self) nogil:
        """The L2 norm of the weight vector. """
        return sqrt(self.sq_norm)

```

</details>

Reply:
So I probably mess-up my minimal example

```python
%%cython --annotate

cdef float x = 32.0
z = 43.0

x < z

cdef float z_float = z
x < z_float
```

<details>
<summary>annotation/summary>

```bash
1: 

+2: cdef float x = 32.0

  __pyx_v_46_cython_magic_01050efeec36b95e006a0231b754feed_x = 32.0;

+3: z = 43.0

  if (PyDict_SetItem(__pyx_d, __pyx_n_s_z, __pyx_float_43_0) < 0) __PYX_ERR(0, 3, __pyx_L1_error)

 4: 

+5: x < z

  __pyx_t_1 = PyFloat_FromDouble(__pyx_v_46_cython_magic_01050efeec36b95e006a0231b754feed_x); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 5, __pyx_L1_error)
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_GetModuleGlobalName(__pyx_t_2, __pyx_n_s_z); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 5, __pyx_L1_error)
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = PyObject_RichCompare(__pyx_t_1, __pyx_t_2, Py_LT); __Pyx_XGOTREF(__pyx_t_3); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 5, __pyx_L1_error)
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

 6: 

+7: cdef float z_float = z

  __Pyx_GetModuleGlobalName(__pyx_t_3, __pyx_n_s_z); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 7, __pyx_L1_error)
  __Pyx_GOTREF(__pyx_t_3);
  __pyx_t_4 = __pyx_PyFloat_AsFloat(__pyx_t_3); if (unlikely((__pyx_t_4 == (float)-1) && PyErr_Occurred())) __PYX_ERR(0, 7, __pyx_L1_error)
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_v_46_cython_magic_01050efeec36b95e006a0231b754feed_z_float = __pyx_t_4;

+8: x < z_float

  (void)((__pyx_v_46_cython_magic_01050efeec36b95e006a0231b754feed_x < __pyx_v_46_cython_magic_01050efeec36b95e006a0231b754feed_z_float));
```

</details>

Reply:
Basically, we are templating so we don't stock the value in Python variable. So @ogrisel would be even better now :) @mbatoul you should accept the proposal :)